File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/SemanticEquivalenceProof.java
package com.foundgine.core.semantic.planning;

import java.util.Objects;

/**
 * Records that a rewrite preserved the provider-neutral semantic meaning of a
 * plan.
 */
public record SemanticEquivalenceProof(String beforeFingerprint, String afterFingerprint) {
	public boolean isSatisfied() {
		return Objects.equals(beforeFingerprint, afterFingerprint);
	}

	public static SemanticEquivalenceProof create(SemanticPlan before, SemanticPlan after) {
		Objects.requireNonNull(before, "before");
		Objects.requireNonNull(after, "after");
		var proof = new SemanticEquivalenceProof(SemanticEquivalenceFingerprint.create(before),
				SemanticEquivalenceFingerprint.create(after));
		if (!proof.isSatisfied())
			throw new IllegalStateException(
					"Semantic rewrite rejected because the rewritten plan does not preserve the canonical semantic meaning of the source plan.");
		return proof;
	}
}

File: src/csharp/Foundgine.Core/Semantic/Planning/SemanticEquivalenceProof.cs
namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Records that a rewrite preserved the provider-neutral semantic meaning of
/// a plan. This is an equivalence check over Foundgine's canonical semantic
/// representation; it is not a proof that a provider implementation is bug-free.
/// </summary>
public sealed record SemanticEquivalenceProof(
    string BeforeFingerprint,
    string AfterFingerprint)
{
    public bool IsSatisfied => string.Equals(BeforeFingerprint, AfterFingerprint, StringComparison.Ordinal);

    public static SemanticEquivalenceProof Create(SemanticPlan before, SemanticPlan after)
    {
        ArgumentNullException.ThrowIfNull(before);
        ArgumentNullException.ThrowIfNull(after);

        var proof = new SemanticEquivalenceProof(
            SemanticEquivalenceFingerprint.Create(before),
            SemanticEquivalenceFingerprint.Create(after));

        if (!proof.IsSatisfied)
            throw new InvalidOperationException(
                "Semantic rewrite rejected because the rewritten plan does not preserve the canonical semantic meaning of the source plan.");

        return proof;
    }
}
